design/client-initiated-backchannel-authentication-flow.md

```diff
@@ -0,0 +1,494 @@
+
+* scope                : specified by Backchannel Authentication Request's "scope" parameter
+* client default scope : set by keycloak Admin Console's Client Settings
+
+#### Binding Message
+
+On CIBA Protocol specification Section [7.1. Authentication Request](https://openid.net/specs/openid-client-initiated-backchannel-authentication-core-1_0.html#auth_request), Binding Message is supported.
+
+#### Pairwise Identifiers
+
+On CIBA Protocol specification Section [4. Registration and Discovery Metadata](https://openid.net/specs/openid-client-initiated-backchannel-authentication-core-1_0.html#registration), Poll Modes with Pairwise Identifiers is NOT supported.
+
+#### User Code
+
+On CIBA Protocol specification Section [7.1. Authentication Request](https://openid.net/specs/openid-client-initiated-backchannel-authentication-core-1_0.html#auth_request), User Code is NOT supported.
+
+#### Requested Expiry
+
+On CIBA Protocol specification Section [7.1. Authentication Request](https://openid.net/specs/openid-client-initiated-backchannel-authentication-core-1_0.html#auth_request), Requested Expiry is NOT supported.
+
+#### Signed Authentication Request
+
+On CIBA Protocol specification Section [7.1.1. Signed Authentication Request](https://openid.net/specs/openid-client-initiated-backchannel-authentication-core-1_0.html#signed_auth_request), Signed Authentication Request is NOT supported.
+
+### Outside CIBA Protocol
+
+The process of AuthN/AuthZ by AD is highly use-cases dependent (e.g. Push notification to the end-user's smart device, SMS message to the end-user's smart device, etc.).
+
+To keep the design in use-cases independent, this design document describes the interface between inside and outside CIBA protocol.
+
+However, if keycloak lacks the actual implementation of these interfaces, the user of keycloak does not use CIBA flow unless they implement these interfaces by themselves. It is inconvenient for users.
+
+Therefore, this design document also defines the default implementation for these interfaces.
+
+### Runtime Environment
+
+This version of the design document considers both standalone and clustering environments but not consider multi-cluster environment.
+
+### Settings
+
+To configure inside CIBA part's operation, CIBA policy is introduced as part of Admin Console's Realm Settings.
+
+#### CIBA Policy
+
+The following items defined in CIBA protocol specification can be configured per realm :
+
+* backchannel_token_delivery_mode : fix "poll"
+* expires_in
+* interval : if set to 0, the token request throttling is deactivated
+
+### Single Sign On
+
+keycloak cannot realize SSO by CIBA flow.
+
+To realize SSO, keycloak uses Cookies retained in the end-user's browser. In CIBA protocol, the end-user's browser does not appear in the CIBA flow.
+
+## Design and Implementation
+
+This section describes how to realize what Specification section determined.
```

Comment:
Indeed, the `auth_req_id` works as a code. I think this approach fits. That is a good catch.

In regards to its value, if I read the specs correctly it is opaque and we can have whatever we want.

Reply:
In regards to the login hint resolver, that is something I was thinking too when I did this comment https://github.com/keycloak/keycloak-community/pull/105#discussion_r431039174.

So we could use the `login_hint_token` for this. If I read the spec correctly, it is all about having some agreement about how this token will look like.

Reply:
I am also not sure if "auth_req_id" needs to be JWE or signed JWT. The "auth_req_id" will contain the AuthResultID in it, this is probably fine to expose to the client?

Reply:
@pedroigor @mposolda Considering this discussion, what do you think about the following?

1.  When a backchannel authentication arrives, keycloak generates `Auth Result ID` and also generate `auth_req_id` which includes parameters that are held by the cache `Auth Req Context Cache` I've initially proposed.

1.  `auth_req_id` is signed and encrypted by JWS/JWE, which make unnecessary `Auth Req Context Cache` I've initially proposed.

1.  `Auth Result Cache` can be realized by the existing `actionToken` cache via the provider refactoring `CodeToTokenStoreProvider`. `Auth Result Cache` is for single-use, but considering the following case that the CD(client) requests tokens before the completion of authentication by AD so that keycloak only removes this cache entry after the completion of authentication by  AD.

1.  When Token Request arrives, keycloak at first check whether access throttling is applied or not. If not, keycloak decrypt the `auth_req_id`, extract the key for accessing the cache entry of `Auth Result Cache`, get and remove this entry to find out the result of authentication by AD.

Reply:
+1 I'm just not sure about the need for encryption but only signing.

Reply:
+1 . Also +1 to @pedroigor 's comment about encryption VS signing. It seems like a trade-off etween better security (using encryption+signing) and less complexity (using just signing). Encryption is needed just in case that we don't want client to know values of some parameters? Leaving to your judge if this would be an issue or not...

Reply:
@mposolda At this time I'd like to employ JWS. However I'd like to consider JWE as an option. I'm afraid that something need to be secret might arise when I start the implementation.